src/QMCWaveFunctions/BsplineFactory/SplineC2COMP.cpp

```diff
@@ -0,0 +1,1294 @@
+    offload_scratch.resize(padded_size * nVP);
+  const auto orb_size = psiinv.size();
+  if (results_scratch.size() < orb_size * nVP)
+    results_scratch.resize(orb_size * nVP);
+
+  // Ye: need to extract sizes and pointers before entering target region
+  const auto* spline_ptr         = SplineInst->getSplinePtr();
+  auto* offload_scratch_ptr      = offload_scratch.data();
+  auto* results_scratch_ptr      = results_scratch.data();
+  const auto myKcart_padded_size = myKcart->capacity();
+  auto* myKcart_ptr              = myKcart->data();
+  auto* psiinv_ptr               = psiinv_pos_copy.data();
+  auto* ratios_private_ptr       = ratios_private.data();
+  const size_t first_spo_local   = first_spo;
+
+  {
+    ScopedTimer offload(&offload_timer_);
+    PRAGMA_OFFLOAD("omp target teams distribute collapse(2) num_teams(NumTeams*nVP) \
+                map(always, to: psiinv_ptr[0:psiinv_pos_copy.size()]) \
+                map(always, from: ratios_private_ptr[0:NumTeams*nVP])")
+    for (int iat = 0; iat < nVP; iat++)
+      for (int team_id = 0; team_id < NumTeams; team_id++)
+      {
+        const int first = ChunkSizePerTeam * team_id;
+        const int last  = (first + ChunkSizePerTeam) > padded_size ? padded_size : first + ChunkSizePerTeam;
+        auto* restrict offload_scratch_iat_ptr = offload_scratch_ptr + padded_size * iat;
+        auto* restrict psi_iat_ptr             = results_scratch_ptr + orb_size * iat;
+        auto* restrict pos_scratch             = reinterpret_cast<RealType*>(psiinv_ptr + orb_size);
+
+        int ix, iy, iz;
+        ST a[4], b[4], c[4];
+        spline2::computeLocationAndFractional(spline_ptr, ST(pos_scratch[iat * 6 + 3]), ST(pos_scratch[iat * 6 + 4]),
+                                              ST(pos_scratch[iat * 6 + 5]), ix, iy, iz, a, b, c);
+
+        PRAGMA_OFFLOAD("omp parallel for")
+        for (int index = 0; index < last - first; index++)
+          spline2offload::evaluate_v_impl_v2(spline_ptr, ix, iy, iz, a, b, c, offload_scratch_iat_ptr + first, first,
+                                             index);
+        const int last_index = last / 2 < orb_size ? last / 2 : orb_size;
+        PRAGMA_OFFLOAD("omp parallel for")
+        for (int index = first / 2; index < last_index; index++)
+          C2C::assign_v(ST(pos_scratch[iat * 6]), ST(pos_scratch[iat * 6 + 1]), ST(pos_scratch[iat * 6 + 2]),
+                        psi_iat_ptr, orb_size, offload_scratch_iat_ptr, myKcart_ptr, myKcart_padded_size,
+                        first_spo_local, index);
+
+        const int first_cplx = first / 2;
+        const int last_cplx  = orb_size < last / 2 ? orb_size : last / 2;
+        ComplexT sum(0);
+        PRAGMA_OFFLOAD("omp parallel for simd reduction(+:sum)")
+        for (int i = first_cplx; i < last_cplx; i++)
+          sum += psi_iat_ptr[i] * psiinv_ptr[i];
+        ratios_private_ptr[iat * NumTeams + team_id] = sum;
+      }
+  }
+
+  // do the reduction manually
```

Comment:
Why? (I think we might be evading a compiler weakness. Under what conditions do we revisit?)

Reply:
GPU reduction across SM/CU is complicated. So done on the host.